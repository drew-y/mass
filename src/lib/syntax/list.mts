import { Expr } from "./expr.mjs";
import { isList } from "./helpers.mjs";
import { Identifier } from "./identifier.mjs";
import { Int } from "./int.mjs";
import { Syntax, SyntaxOpts, SyntaxComparable } from "./syntax.mjs";

export class List extends Syntax {
  readonly __type = "list";
  value: Expr[] = [];

  constructor(opts: SyntaxOpts & { value?: ListValue[] }) {
    super(opts);
    this.push(...(opts.value ?? []));
  }

  get hasChildren() {
    return !!this.value.length;
  }

  get length() {
    return new Int({ value: this.value.length });
  }

  at(index: number): Expr | undefined {
    return this.value.at(index);
  }

  calls(fnId: Expr | string) {
    return !!this.at(0)?.is(fnId);
  }

  consume(): Expr {
    const next = this.value.shift();
    if (!next) {
      throw new Error("No remaining expressions");
    }
    return next;
  }

  first(): Expr | undefined {
    return this.value[0];
  }

  rest(): Expr[] {
    return this.value.slice(1);
  }

  pop(): Expr | undefined {
    return this.value.pop();
  }

  push(...expr: ListValue[]) {
    expr.forEach((ex) => {
      if (typeof ex === "string") {
        this.value.push(new Identifier({ value: ex, parent: this }));
        return;
      }

      if (ex instanceof Array) {
        this.push(new List({ value: ex, parent: this }));
        return;
      }

      const cloned = ex.clone(this);

      if (isList(cloned) && cloned.calls("splice-block")) {
        this.value.push(...cloned.rest());
        return;
      }

      this.value.push(cloned);
    });

    return this;
  }

  indexOf(expr: Expr) {
    return this.value.findIndex((v) => v.is(expr));
  }

  insert(expr: Expr | string, at = 0) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    result.setParent(this);
    this.value.splice(at, 0, result);
    return this;
  }

  is(_?: SyntaxComparable): boolean {
    return false;
  }

  map(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    return new List({ value: this.value.map(fn), from: this });
  }

  reduce(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    const list = new List({ from: this });
    return this.value.reduce((newList: List, expr, index, array) => {
      if (!expr) return newList;
      return newList.push(fn(expr, index, array));
    }, list);
  }

  slice(start?: number, end?: number): List {
    return new List({ from: this, value: this.value.slice(start, end) });
  }

  toJSON() {
    return this.value;
  }

  clone(parent?: Expr): List {
    return new List({ parent, value: this.value, from: this });
  }
}

export type ListValue = Expr | string | ListValue[];
