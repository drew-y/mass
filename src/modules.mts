import { ParsedFiles } from "./lib/parse-directory.mjs";
import { List } from "./syntax-objects/list.mjs";
import { VoidModule } from "./syntax-objects/module.mjs";

export const resolveFileModules = (opts: {
  /** Path to the std lib directory */
  stdPath: string;
  /** Path to the user source code */
  srcPath: string;
  files: ParsedFiles;
}): VoidModule => {
  const { stdPath, srcPath, files } = opts;

  const rootModule = new VoidModule({ name: "root" });

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(filePath, srcPath, stdPath);
    const parsedPath = resolvedPath.split("/").filter(Boolean);
    registerModule({ path: parsedPath, parentModule: rootModule, ast: file });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
}: {
  path: string[];
  parentModule: VoidModule;
  ast: List;
}): VoidModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveChildEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && name === "index") {
    parentModule.push(...ast.value);
    return;
  }

  const module =
    existingModule ??
    new VoidModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.value } : {}),
      name,
    });
  module.isExported = true;

  if (!existingModule) parentModule.push(module);

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (
  filePath: string,
  srcPath: string,
  stdPath: string
) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".void", "");

  return finalPath;
};
