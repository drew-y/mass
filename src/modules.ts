import { ParsedFiles } from "./lib/parse-directory.js";
import { List } from "./syntax-objects/list.js";
import { VoidModule } from "./syntax-objects/module.js";

export const resolveFileModules = (opts: {
  /** Path to the std lib directory */
  stdPath: string;
  /** Path to the user source code root folder */
  srcPath?: string;
  /** Path to the entry index file, which determines what is exported from the package */
  indexPath: string;
  files: ParsedFiles;
}): VoidModule => {
  const { stdPath, srcPath, files } = opts;

  const rootModule = new VoidModule({ name: "root" });

  for (const [filePath, file] of Object.entries(files)) {
    const resolvedPath = filePathToModulePath(
      filePath,
      srcPath ?? opts.indexPath,
      stdPath
    );

    const parsedPath = resolvedPath.split("/").filter(Boolean);

    registerModule({
      path: parsedPath,
      parentModule: rootModule,
      ast: file.slice(1), // Skip the first element (ast)
      isIndex: filePath === opts.indexPath,
    });
  }

  return rootModule;
};

const registerModule = ({
  path,
  parentModule,
  ast,
  isIndex,
}: {
  path: string[];
  parentModule: VoidModule;
  ast: List;
  isIndex?: boolean;
}): VoidModule | undefined => {
  const [name, ...rest] = path;

  if (!name) return;

  const existingModule = parentModule.resolveChildEntity(name);

  if (existingModule && !existingModule.isModule()) {
    throw new Error(
      `Cannot register module ${name} because it is already registered as ${existingModule.syntaxType}`
    );
  }

  if (!existingModule && name === "index") {
    parentModule.push(...ast.toArray());
    return;
  }

  const module =
    existingModule ??
    new VoidModule({
      ...(!rest.length ? { ...ast.metadata, value: ast.toArray() } : {}),
      name,
      isIndex,
    });
  module.isExported = true;

  if (!existingModule) parentModule.push(module);

  if (!rest.length) return;

  return registerModule({ path: rest, parentModule: module, ast });
};

const filePathToModulePath = (
  filePath: string,
  srcPath: string,
  stdPath: string
) => {
  let finalPath = filePath.startsWith(stdPath)
    ? filePath.replace(stdPath, "std")
    : filePath;

  finalPath = finalPath.startsWith(srcPath)
    ? finalPath.replace(srcPath, "src")
    : finalPath;

  finalPath = finalPath.replace(".void", "");

  return finalPath;
};
