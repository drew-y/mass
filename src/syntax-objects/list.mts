import { Expr } from "./expr.mjs";
import { getIdStr } from "./get-id-str.mjs";
import { Id, Identifier } from "./identifier.mjs";
import { Int } from "./int.mjs";
import { Syntax, SyntaxOpts } from "./syntax.mjs";

export class List extends Syntax {
  readonly syntaxType = "list";
  value: Expr[] = [];

  constructor(
    opts: SyntaxOpts & {
      value?: ListValue[] | List;
    }
  ) {
    super(opts);
    const value = opts.value;

    if (!value || value instanceof Array) {
      this.push(...(value ?? []));
    } else {
      this.push(...value.value);
    }
  }

  get hasChildren() {
    return !!this.value.length;
  }

  get length() {
    return new Int({ value: this.value.length });
  }

  at(index: number): Expr | undefined {
    return this.value.at(index);
  }

  identifierAt(index: number): Identifier {
    const id = this.at(index);
    if (!id?.isIdentifier()) {
      throw new Error(`No identifier at index ${index}`);
    }
    return id;
  }

  listAt(index: number): List {
    const id = this.at(index);
    if (!id?.isList()) {
      throw new Error(`No list at index ${index}`);
    }
    return id;
  }

  set(index: number, value: Expr) {
    value.parent = this;
    this.value[index] = value; // Should this clone?
    return this;
  }

  calls(fnId: Id, atIndex = 0) {
    return this.getIdStrAt(atIndex) === getIdStr(fnId);
  }

  getIdStrAt(index: number): string | undefined {
    const v = this.at(index);
    return v?.isIdentifier() || v?.isStringLiteral() ? v.value : undefined;
  }

  consume(): Expr {
    const next = this.value.shift();
    if (!next) throw new Error("No remaining expressions");
    return next;
  }

  first(): Expr | undefined {
    return this.value[0];
  }

  /** Returns all but the first element in an array */
  rest(): Expr[] {
    return this.value.slice(1);
  }

  pop(): Expr | undefined {
    return this.value.pop();
  }

  push(...expr: ListValue[]) {
    expr.forEach((ex) => {
      if (typeof ex === "string") {
        this.value.push(new Identifier({ value: ex, parent: this }));
        return;
      }

      if (ex instanceof Array) {
        this.push(new List({ value: ex, parent: this }));
        return;
      }

      const cloned = ex.clone(this);

      if (cloned.isList() && cloned.calls("splice-quote")) {
        this.value.push(...cloned.rest());
        return;
      }

      this.value.push(cloned);
    });

    return this;
  }

  findIndex(cb: (expr: Expr) => boolean) {
    return this.value.findIndex(cb);
  }

  insert(expr: Expr | string, at = 0) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    result.parent = this;
    this.value.splice(at, 0, result);
    return this;
  }

  filter(fn: (expr: Expr, index: number, array: Expr[]) => boolean): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.value.filter(fn),
    });
  }

  map(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.value.map(fn),
    });
  }

  /** Returns a copy of this list where all the parameters mapped by the supplied function */
  mapArgs(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    const newList = new List({
      ...super.getCloneOpts(),
      value: this.rest().map(fn),
    });
    if (this.first()) newList.insert(this.first()!);
    return newList;
  }

  reduce(
    fn: (expr: Expr, index: number, array: Expr[]) => Expr | undefined
  ): List {
    const list = new List({ ...super.getCloneOpts() });
    return this.value.reduce((newList: List, expr, index, array) => {
      if (!expr) return newList;
      const result = fn(expr, index, array);
      if (!result) return newList;
      return newList.push(result);
    }, list);
  }

  slice(start?: number, end?: number): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.value.slice(start, end),
    });
  }

  toJSON() {
    return this.value;
  }

  clone(parent?: Expr): List {
    return new List({ ...super.getCloneOpts(parent), value: this });
  }
}

export type ListValue = Expr | string | ListValue[];
