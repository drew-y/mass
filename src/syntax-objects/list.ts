import { Expr } from "./expr.js";
import { Float } from "./float.js";
import { getIdStr } from "./lib/get-id-str.js";
import { Id, Identifier } from "./identifier.js";
import { Int } from "./int.js";
import { Syntax, SyntaxMetadata } from "./syntax.js";
import { ChildList } from "./lib/child-list.js";

type ListOpts =
  | ListValue[]
  | (SyntaxMetadata & {
      value?: ListValue[] | List | ChildList<Expr>;
      isParentheticalList?: boolean;
    });

export class List extends Syntax {
  readonly syntaxType = "list";
  /** True when the list was defined by the user using parenthesis i.e. (hey, there) */
  mayBeTuple?: boolean;
  #store = new ChildList(undefined, this);

  constructor(opts: ListOpts) {
    opts = Array.isArray(opts) ? { value: opts } : opts;
    super(opts);
    const value = opts.value;
    this.mayBeTuple = opts.isParentheticalList;

    if (!value || value instanceof Array) {
      this.push(...(value ?? []));
    } else if (value instanceof List) {
      this.push(...value.toArray());
    }
  }

  get children() {
    return this.toArray();
  }

  get hasChildren() {
    return !!this.#store.length;
  }

  get length() {
    return this.#store.length;
  }

  at(index: number): Expr | undefined {
    return this.#store.at(index);
  }

  exprAt(index: number): Expr {
    const expr = this.#store.at(index);
    if (!expr) {
      throw new Error(`No expr at ${index}`);
    }
    return expr;
  }

  identifierAt(index: number): Identifier {
    const id = this.at(index);
    if (!id?.isIdentifier()) {
      throw new Error(`No identifier at index ${index}`);
    }
    return id;
  }

  optionalIdentifierAt(index: number): Identifier | undefined {
    const id = this.at(index);
    if (id?.isIdentifier()) {
      return id;
    }
  }

  listAt(index: number): List {
    const id = this.at(index);
    if (!id?.isList()) {
      throw new Error(`No list at index ${index}`);
    }
    return id;
  }

  set(index: number, expr: Expr | string) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.set(index, result);
    return this;
  }

  calls(fnId: Id, atIndex = 0) {
    return this.getIdStrAt(atIndex) === getIdStr(fnId);
  }

  getIdStrAt(index: number): string | undefined {
    const v = this.at(index);
    return v?.isIdentifier() || v?.isStringLiteral() ? v.value : undefined;
  }

  consume(): Expr {
    return this.#store.consume();
  }

  first(): Expr | undefined {
    return this.#store.at(0);
  }

  last(): Expr | undefined {
    return this.#store.at(-1);
  }

  /** Returns all but the first element in an array */
  argsArray(): Expr[] {
    return this.#store.toArray().slice(1);
  }

  pop(): Expr | undefined {
    return this.#store.pop();
  }

  push(...expr: ListValue[]) {
    expr.forEach((ex) => {
      if (typeof ex === "string") {
        this.#store.push(new Identifier({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number" && Number.isInteger(ex)) {
        this.#store.push(new Int({ value: ex, parent: this }));
        return;
      }

      if (typeof ex === "number") {
        this.#store.push(new Float({ value: ex, parent: this }));
        return;
      }

      if (ex instanceof Array) {
        this.push(new List({ value: ex, parent: this }));
        return;
      }

      if (ex.isList() && ex.calls("splice_quote")) {
        this.#store.push(...ex.argsArray());
        return;
      }

      this.#store.push(ex);
    });

    return this;
  }

  findIndex(cb: (expr: Expr) => boolean) {
    return this.toArray().findIndex(cb);
  }

  insert(expr: Expr | string, at = 0) {
    const result = typeof expr === "string" ? Identifier.from(expr) : expr;
    this.#store.insert(result, at);
    return this;
  }

  remove(index: number, count = 1) {
    this.#store.remove(index, count);
    return this;
  }

  filter(fn: (expr: Expr, index: number, array: Expr[]) => boolean): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().filter(fn),
    });
  }

  each(fn: (expr: Expr, index: number, array: Expr[]) => void): List {
    this.toArray().forEach(fn);
    return this;
  }

  map(fn: (expr: Expr, index: number, array: Expr[]) => Expr): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.toArray().map(fn),
    });
  }

  /** Like a regular map, but omits undefined values returned from the mapper */
  mapFilter(
    fn: (expr: Expr, index: number, array: Expr[]) => Expr | undefined
  ): List {
    const list = new List({ ...super.getCloneOpts() });
    return this.toArray().reduce((newList: List, expr, index, array) => {
      if (!expr) return newList;
      const result = fn(expr, index, array);
      if (!result) return newList;
      return newList.push(result);
    }, list);
  }

  slice(start?: number, end?: number): List {
    return new List({
      ...super.getCloneOpts(),
      value: this.#store.slice(start, end),
    });
  }

  sliceAsArray(start?: number, end?: number) {
    return this.children.slice(start, end);
  }

  toArray(): Expr[] {
    return this.#store.toArray();
  }

  toJSON() {
    return this.toArray();
  }

  /** Clones should return a deep copy of all expressions except for type expressions */
  clone(parent?: Expr): List {
    return new List({
      ...super.getCloneOpts(parent),
      value: this.#store.toClonedArray(),
      isParentheticalList: this.mayBeTuple,
    });
  }
}

export type ListValue = Expr | string | number | ListValue[];
