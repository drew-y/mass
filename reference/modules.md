## Modules

```void
use src::lib::* // Import everything from src/lib
use src::lib as my-lib // Import everything as my-lib
use src::lib::{ my-util-function } // Import my-util-function from src/lib
use src::lib::{ func-a sub-module: { func-b } } // Import func-a from src/lib and func-b from the submodule in src/lib
use super::helpers::{ func-a as func-c } // Import func-a as func-c from the parent module of helpers
// If the path points to a folder, an index.void is assumed
use src::folder::{ b } // Resolves to src/folder/index.void
use package::my_package::{ pack-func } // Import pack-func from the installed package called my_package.

mod my-module // Create a new module within the current module
  fn my-func() // Define a function within the module
    print "Hello from my-module"
```
