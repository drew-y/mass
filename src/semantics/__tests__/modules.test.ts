import { registerModules } from "../modules.js";
import { List } from "../../syntax-objects/list.js";
import { stdPath } from "../../parser/index.js";
import { test } from "vitest";

test("module registration", (t) => {
  const result = registerModules(input);
  t.expect(result).toMatchSnapshot();
});

const input = {
  files: {
    "/Users/drew/projects/void/example.void": new List({
      value: [
        "ast",
        ["use", ["::", ["::", "std", "macros"], "all"]],
        [
          "fn",
          ["fib", [":", "n", "i32"]],
          "->",
          "i32",
          [
            "block",
            [
              "if",
              ["<=", "n", 1],
              [":", "then", ["block", "n"]],
              [
                ":",
                "else",
                [
                  "block",
                  ["+", ["fib", ["-", "n", 1]], ["fib", ["-", "n", 2]]],
                ],
              ],
            ],
          ],
        ],
        [
          "fn",
          ["main"],
          [
            "block",
            ["let", ["=", "x", ["+", 10, ["block", ["+", 20, 30]]]]],
            [
              "let",
              [
                "=",
                "y",
                [
                  "if",
                  [">", "x", 10],
                  [":", "then", ["block", 10]],
                  [":", "else", ["block", 20]],
                ],
              ],
            ],
            [
              "call",
              "this",
              "while",
              [
                "=>",
                [],
                [
                  "if",
                  [">", "x", 10],
                  [":", "then", ["block", ["-=", "x", 1]]],
                  [":", "else", ["block", ["+=", "x", 1]]],
                ],
              ],
            ],
            [
              "let",
              [
                "=",
                "n",
                [
                  "if",
                  [">", ["len", "args"], 1],
                  [
                    ":",
                    "then",
                    [
                      "block",
                      ["log", "console", ["string", "Hey there!"]],
                      ["unwrap", ["parseInt", ["at", "args", 1]]],
                    ],
                  ],
                  [":", "else", ["block", 10]],
                ],
              ],
            ],
            ["let", ["=", "x2", 10]],
            ["let", ["=", "z", ["nothing"]]],
            ["let", ["=", "test_spacing", ["fib", "n"]]],
            ["let", ["=", "result", ["fib", "n"]]],
          ],
        ],
      ],
    }),
    [`${stdPath}/memory.void`]: new List({
      value: [
        "ast",
        ["use", ["::", ["::", "super", "macros"], "all"]],
        ["global", "let", ["=", "header-size", 8]],
        ["global", "let", ["=", "size-index", 0]],
        ["global", "let", ["=", "type-index", 4]],
        ["global", "var", ["=", "stack-pointer", 0]],
        [
          "pub",
          "fn",
          ["copy", [":", "src", "i32"], [":", "dest", "i32"]],
          "->",
          "i32",
          [
            "block",
            [
              "bnr",
              ["memory", "copy", "void"],
              ["dest", "src", ["size", "src"]],
            ],
            "dest",
          ],
        ],
      ],
    }),
    [`${stdPath}/index.void`]: new List({
      value: ["ast", ["pub", ["use", ["::", "macros", "all"]]]],
    }),
  },
  srcPath: "/Users/drew/projects/void",
  indexPath: "/Users/drew/projects/void/index.void",
};
