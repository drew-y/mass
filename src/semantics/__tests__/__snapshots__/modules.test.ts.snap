// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module registration 1`] = `
[
  "module",
  "root",
  [
    "exports",
    [
      "src#291",
      "std#308",
    ],
  ],
  [
    [
      "module",
      "src",
      [
        "exports",
        [],
      ],
      [
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "module",
          "example",
          [
            "exports",
            [],
          ],
          [
            [
              "use",
              [
                "::",
                "root",
                "all",
              ],
            ],
            [
              "use",
              [
                "::",
                [
                  "::",
                  "std",
                  "macros",
                ],
                "all",
              ],
            ],
            [
              "fn",
              [
                "fib",
                [
                  ":",
                  "n",
                  "i32",
                ],
              ],
              "->",
              "i32",
              [
                "block",
                [
                  "if",
                  [
                    "<=",
                    "n",
                    1,
                  ],
                  [
                    ":",
                    "then",
                    [
                      "block",
                      "n",
                    ],
                  ],
                  [
                    ":",
                    "else",
                    [
                      "block",
                      [
                        "+",
                        [
                          "fib",
                          [
                            "-",
                            "n",
                            1,
                          ],
                        ],
                        [
                          "fib",
                          [
                            "-",
                            "n",
                            2,
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
              ],
            ],
            [
              "fn",
              [
                "main",
              ],
              [
                "block",
                [
                  "let",
                  [
                    "=",
                    "x",
                    [
                      "+",
                      10,
                      [
                        "block",
                        [
                          "+",
                          20,
                          30,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "y",
                    [
                      "if",
                      [
                        ">",
                        "x",
                        10,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          10,
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          20,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "call",
                  "this",
                  "while",
                  [
                    "=>",
                    [],
                    [
                      "if",
                      [
                        ">",
                        "x",
                        10,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          [
                            "-=",
                            "x",
                            1,
                          ],
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          [
                            "+=",
                            "x",
                            1,
                          ],
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "n",
                    [
                      "if",
                      [
                        ">",
                        [
                          "len",
                          "args",
                        ],
                        1,
                      ],
                      [
                        ":",
                        "then",
                        [
                          "block",
                          [
                            "log",
                            "console",
                            [
                              "string",
                              "Hey there!",
                            ],
                          ],
                          [
                            "unwrap",
                            [
                              "parseInt",
                              [
                                "at",
                                "args",
                                1,
                              ],
                            ],
                          ],
                        ],
                      ],
                      [
                        ":",
                        "else",
                        [
                          "block",
                          10,
                        ],
                      ],
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "x2",
                    10,
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "z",
                    [
                      "nothing",
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "test_spacing",
                    [
                      "fib",
                      "n",
                    ],
                  ],
                ],
                [
                  "let",
                  [
                    "=",
                    "result",
                    [
                      "fib",
                      "n",
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    [
      "module",
      "std",
      [
        "exports",
        [],
      ],
      [
        [
          "pub",
          [
            "use",
            [
              "::",
              "macros",
              "all",
            ],
          ],
        ],
        [
          "use",
          [
            "::",
            "root",
            "all",
          ],
        ],
        [
          "module",
          "memory",
          [
            "exports",
            [],
          ],
          [
            [
              "use",
              [
                "::",
                "root",
                "all",
              ],
            ],
            [
              "use",
              [
                "::",
                [
                  "::",
                  "super",
                  "macros",
                ],
                "all",
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "header-size",
                8,
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "size-index",
                0,
              ],
            ],
            [
              "global",
              "let",
              [
                "=",
                "type-index",
                4,
              ],
            ],
            [
              "global",
              "var",
              [
                "=",
                "stack-pointer",
                0,
              ],
            ],
            [
              "pub",
              "fn",
              [
                "copy",
                [
                  ":",
                  "src",
                  "i32",
                ],
                [
                  ":",
                  "dest",
                  "i32",
                ],
              ],
              "->",
              "i32",
              [
                "block",
                [
                  "bnr",
                  [
                    "memory",
                    "copy",
                    "void",
                  ],
                  [
                    "dest",
                    "src",
                    [
                      "size",
                      "src",
                    ],
                  ],
                ],
                "dest",
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;
